### ISSUE TITLE
System requirements for generating planet tiles for zoom 1-5

### ISSUE THREAD

**Issue by kuwapa** [systemed/tilemaker#322]
Hi, I want to generate mbtiles file for the planet only for zoom 1-5. Would a machine with 64GB RAM be enough for this or do I need to use a 144GB machine as you did?

>  @systemed - Yes, you should be ok with 64GB if you filter the planet file first to remove anything that you’re not going to need (which if it’s z1-5 I guess is most things). You could possibly just use Natural Earth shapefiles for a z1-5 map and avoid using OSM altogether.

By filter, you mean with a tool like osmium? 
I would still need the country names and roads and such, for that natural earth shapefiles wouldnt work right?

> @systemed - Yes - osmium or osmfilter (osmium’s probably easier). It depends what detail you need - Natural Earth does have a major roads shapefile, for example. Basically you want to avoid reading anything into memory that you don’t have to!

**Comment by kuwapa** [comment]
I looked into `osmium` and `osmfilter`. Unfortunately both tools do not seem to have any methods to filter by zoom levels. The only have the ability to filter by tags, types, relations etc. 

To convert the natural earth shapefile to mbtiles, I would have to use tippecanoe, correct?

I've almost finished download the planet.osm.pbf file on a 36 core, 72GB RAM EC2 instance on AWS and am gonna give it a shot to generate a MBTiles file for zoom 1-5 using tilemaker by changing the config to 

```
"settings": {
		"minzoom": 0,
		"maxzoom": 5,
		"basezoom": 5,
		"include_ids": false,
		"combine_below": 14,
		"name": "Tilemaker to OpenMapTiles schema",
		"version": "3.0",
		"description": "Tile config based on OpenMapTiles schema",
		"compress": "gzip",
		"filemetadata": {
			"tilejson": "2.0.0", 
			"scheme": "xyz", 
			"type": "baselayer", 
			"format": "pbf", 
            "tiles": ["https://example.com/liechtenstein/{z}/{x}/{y}.pbf"]
		}
	}
```

This would work right?

**Comment by kuwapa** [comment]
Unfortunately, the process failed. These are the logs - 

> tilemaker --input planet-latest.osm.pbf --output planet.mbtiles --config new_config.json --process new-process-openmaptiles.lua 

```
Layer place (z0-14)
Layer boundary (z0-14)
Layer poi (z12-14)
Layer poi_detail (z14-14) -> poi
Layer housenumber (z14-14)
Layer waterway (z8-14)
Layer waterway_detail (z12-14) -> waterway
Layer transportation (z4-14)
Layer transportation_name (z8-14)
Layer transportation_name_mid (z13-14) -> transportation_name
Layer transportation_name_detail (z14-14) -> transportation_name
Layer building (z14-14)
Layer water (z6-14)
Layer ocean (z0-14) -> water
Layer water_name (z14-14)
Layer water_name_detail (z14-14) -> water_name
Layer aeroway (z11-14)
Layer aerodrome_label (z10-14)
Layer park (z11-14)
Layer landuse (z4-14)
Layer urban_areas (z4-8) -> landuse
Layer landcover (z8-14)
Layer ice_shelf (z0-9) -> landcover
Layer glacier (z2-9) -> landcover
Layer mountain_peak (z11-14)
Bounding box -180, -85.06, 180, 85.06
Reading .shp ocean
Shapefile entity #1974 type 5 is invalid. Parts:232. Reason:Geometry has invalid self-intersections. A self-intersection point was found at (-146.579, -15.9312); method: i; operations: u/i; segment IDs {source, multi, ring, segment}: {0, 0, 47, 1}/{0, 0, 47, 23}... failed to correct. Reason: Geometry has invalid self-intersections. A self-intersection point was found at (-146.579, -15.9312); method: i; operations: u/i; segment IDs {source, multi, ring, segment}: {0, 0, 47, 1}/{0, 0, 47, 23}
Shapefile entity #4900 type 5 is invalid. Parts:27. Reason:Geometry has invalid self-intersections. A self-intersection point was found at (-110.642, 25.7673); method: i; operations: u/i; segment IDs {source, multi, ring, segment}: {0, 0, -1, 215}/{0, 0, -1, 217}... failed to correct. Reason: Geometry has invalid self-intersections. A self-intersection point was found at (-110.642, 25.7673); method: i; operations: u/i; segment IDs {source, multi, ring, segment}: {0, 0, -1, 215}/{0, 0, -1, 217}
Shapefile entity #14242 type 5 is invalid. Parts:217. Reason:Geometry has invalid self-intersections. A self-intersection point was found at (-71.4807, 45.5092); method: i; operations: i/u; segment IDs {source, multi, ring, segment}: {0, 0, -1, 30085}/{0, 0, -1, 30087}... failed to correct. Reason: Geometry has invalid self-intersections. A self-intersection point was found at (-71.4807, 45.5092); method: i; operations: i/u; segment IDs {source, multi, ring, segment}: {0, 0, -1, 30085}/{0, 0, -1, 30087}
Shapefile entity #14663 type 5 is invalid. Parts:3156. Reason:Geometry has invalid self-intersections. A self-intersection point was found at (9.87713, 83.8271); method: i; operations: i/u; segment IDs {source, multi, ring, segment}: {0, 0, 1628, 7}/{0, 0, 1628, 9}... failed to correct. Reason: Geometry has invalid self-intersections. A self-intersection point was found at (9.87713, 83.8271); method: i; operations: i/u; segment IDs {source, multi, ring, segment}: {0, 0, 1628, 7}/{0, 0, 1628, 9}
Shapefile entity #14813 type 5 is invalid. Parts:1. Reason:Geometry has too few points... corrected
Shapefile entity #14814 type 5 is invalid. Parts:1. Reason:Geometry has too few points... corrected
Shapefile entity #14815 type 5 is invalid. Parts:1. Reason:Geometry has too few points... corrected
Shapefile entity #14816 type 5 is invalid. Parts:1. Reason:Geometry has too few points... corrected
..
..
..
<Many such logs in the middle>
..
..
Shapefile entity #53272 type 5 is invalid. Parts:1. Reason:Geometry has too few points... corrected
Shapefile entity #53273 type 5 is invalid. Parts:1. Reason:Geometry has too few points... corrected
Shapefile entity #53274 type 5 is invalid. Parts:1. Reason:Geometry has too few points... corrected
Reading .shp urban_areas
Unable to open landcover/ne_10m_urban_areas/ne_10m_urban_areas.shp or landcover/ne_10m_urban_areas/ne_10m_urban_areas.SHP.
Reading .shp ice_shelf
Unable to open landcover/ne_10m_antarctic_ice_shelves_polys/ne_10m_antarctic_ice_shelves_polys.shp or landcover/ne_10m_antarctic_ice_shelves_polys/ne_10m_antarctic_ice_shelves_polys.SHP.
Reading .shp glacier
Unable to open landcover/ne_10m_glaciated_areas/ne_10m_glaciated_areas.shp or landcover/ne_10m_glaciated_areas/ne_10m_glaciated_areas.SHP.
Sorting loaded shapes
Reading .pbf planet-latest.osm.pbf
Killed6061/17889 ways 0 relations 0      
```

I removed the water_polygons.shp file and tried again. It unfortunately failed around 6K/17K ways read

```
Layer place (z0-14)
Layer boundary (z0-14)
Layer poi (z12-14)
Layer poi_detail (z14-14) -> poi
Layer housenumber (z14-14)
Layer waterway (z8-14)
Layer waterway_detail (z12-14) -> waterway
Layer transportation (z4-14)
Layer transportation_name (z8-14)
Layer transportation_name_mid (z13-14) -> transportation_name
Layer transportation_name_detail (z14-14) -> transportation_name
Layer building (z14-14)
Layer water (z6-14)
Layer ocean (z0-14) -> water
Layer water_name (z14-14)
Layer water_name_detail (z14-14) -> water_name
Layer aeroway (z11-14)
Layer aerodrome_label (z10-14)
Layer park (z11-14)
Layer landuse (z4-14)
Layer urban_areas (z4-8) -> landuse
Layer landcover (z8-14)
Layer ice_shelf (z0-9) -> landcover
Layer glacier (z2-9) -> landcover
Layer mountain_peak (z11-14)
Bounding box -180, -85.06, 180, 85.06
Reading .shp ocean
Unable to open water-polygons-split-4326/water_polygons.shp or water-polygons-split-4326/water_polygons.SHP.
Reading .shp urban_areas
Unable to open landcover/ne_10m_urban_areas/ne_10m_urban_areas.shp or landcover/ne_10m_urban_areas/ne_10m_urban_areas.SHP.
Reading .shp ice_shelf
Unable to open landcover/ne_10m_antarctic_ice_shelves_polys/ne_10m_antarctic_ice_shelves_polys.shp or landcover/ne_10m_antarctic_ice_shelves_polys/ne_10m_antarctic_ice_shelves_polys.SHP.
Reading .shp glacier
Unable to open landcover/ne_10m_glaciated_areas/ne_10m_glaciated_areas.shp or landcover/ne_10m_glaciated_areas/ne_10m_glaciated_areas.SHP.
Sorting loaded shapes
Reading .pbf planet-latest.osm.pbf
Killed6315/17889 ways 0 relations 0        
```

What am I doing wrong? Is 72GB RAM not enough?

**Comment by keichan34** [comment]
Usually when I get the "Killed" message, it means tilemaker was killed by the OOM killer, which means, yeah, 72GB is not enough. It looks like to process a raw planet.osm.pbf file you'll need [around 130GB of RAM and 270GB storage](https://github.com/systemed/tilemaker/issues/315#issue-994322040)

If you only need z1-5 and you want it to be based on OSM instead of Natural Earth, use osmium to filter not by zoom level (osm.pbf doesn't have the concept of a zoom level -- it's just a bunch of data), but by the features you want to have in your map at z1-5. For example, major roads, country labels, lakes, rivers, etc.

**Comment by kleunen** [comment]
Use the following extract.json
````
{ "extracts": [
{
    "output": "planet_1.osm.pbf",
    "description": "Bounding box specified in array format",
    "bbox": [-180,-75,0,0]
},
{
    "output": "planet_2.osm.pbf",
    "description": "Bounding box specified in array format",
    "bbox": [0,-75,180,0]
},
{
    "output": "planet_3.osm.pbf",
    "description": "Bounding box specified in array format",
    "bbox": [0,0,180,75]
},
{
    "output": "planet_4.osm.pbf",
    "description": "Bounding box specified in array format",
    "bbox": [-180,0,0,75]
}
] }
````

Then run the conversion as follows:
````
osmium extract --config extract.json --set-bounds --strategy=smart planet.osm.pbf
rm -rf planet.osm.pbf

osmium renumber -v planet_1.osm.pbf --output planet_1.renum.osm.pbf
osmium renumber -v planet_2.osm.pbf --output planet_2.renum.osm.pbf
osmium renumber -v planet_3.osm.pbf --output planet_3.renum.osm.pbf
osmium renumber -v planet_4.osm.pbf --output planet_4.renum.osm.pbf
rm -rf planet_1.osm.pbf planet_2.osm.pbf planet_3.osm.pbf planet_4.osm.pbf

tilemaker --input planet_1.renum.osm.pbf --output planet.mbtiles --config new_config.json --process new-process-openmaptiles.lua --merge --compact --store osm_store.dat
tilemaker --input planet_2.renum.osm.pbf --output planet.mbtiles --config new_config.json --process new-process-openmaptiles.lua --merge --compact --store osm_store.dat
tilemaker --input planet_3.renum.osm.pbf --output planet.mbtiles --config new_config.json --process new-process-openmaptiles.lua --merge --compact --store osm_store.dat
tilemaker --input planet_4.renum.osm.pbf --output planet.mbtiles --config new_config.json --process new-process-openmaptiles.lua --merge --compact --store osm_store.dat
````

With a 72GB mem + 72GB swap machine, this should work. Even up to zoom level 14.

**Comment by systemed** [comment]
All good suggestions.

For z0-5 you have just the following showing (using the OMT-compatible schema and style):

- Placenames: continent, country, state, city
- Boundaries: admin level 3-4
- Roads: motorway, trunk
- Coastline from shapefiles
- Urban areas, landcover, ice shelf, glacier from Natural Earth shapefiles

All of this is available in Natural Earth, so by constructing your own JSON config file and Lua file you could remove the need to read an OSM .pbf at all. It would be quite a useful thing to have shipped with tilemaker though I probably won't have time to do it soon.

**Comment by kuwapa** [comment]
Thank you all for your suggestions, I ended up going the hard way - which was downloading the planet-latest.osm.pbf file and creating a filtered version using osmium-tool. 

This is the config I used for osmium tag filter - 

`
osmium tags-filter planet-latest.osm.pbf w/highway=motorway r/boundary=administrative n/place=city,state,country -o planet.osm.pbf
`

This took about ~15 mins and created a ~960MB planet.osm.pbf file

Before firing off the tilemaker, I made 2 changes to the config
1. Setting max zoom, base zoom under `settings` to 5. 
2. Simplifying the transport layer to this - 
`
"transportation":             { "minzoom": 4,  "maxzoom": 14, "simplify_below": 14, "simplify_level": 0.0002, "combine_below" : 6, "simplify_ratio": 5 },
`
The aim was to create roads which looked as smooth as possible with the least amount of detail. This is the closest I got to it. 

Then fired off tilemaker
`
tilemaker --input planet.osm.pbf --output planet.mbtiles --config planet_config.json --process new-process-openmaptiles.lua
`

I also included the water_polygons.shp so total tiles that were encoded were 1133. The whole process took about ~1.5-2 hours which was a lot considering how few tiles there were to be encoded. The final mbtiles file came out to be 9.6 MB. Here is the file - [planet.mbtiles.zip](https://github.com/systemed/tilemaker/files/7190655/planet.mbtiles.zip)


#### About the EC2 instance
I fired up `i3en.6xlarge` instance type which has a config of 24 cores, 192 GB RAM, 2x7500 NVMe SSD which has an hourly rate of 2.7$. I was experimenting and running it for about 4-5 hours so I ended up 10$ or so bill. It turned out to be less than 2.7 * number of hours; not sure why. If I'd known exactly what to do I think the whole process can be done within 3 hours - 1.5 hours to download the planet file and another 1.5 hours for the processing. 

#### Issues and things to improve
1. I struggled a bit with the highway tags. I initially had added `trunk, primary` to the highway tag to be filtered, but that ended up making the roads look too messy, so I stuck with motorway. However, I discovered that motorway exists only in limited countries and eventually trunk, primary need to be included. However, transport layer will have to be more simplified.
2. Reduce density of cities - All places marked as cities are showing up, which is not how I wanted it. I just wanted the major cities to be displayed. Yet to figure out how to do this. 
3. Boundaries are incomplete in many places - between countries (for eg between Sudan and south Sudan) and even state boundaries (for eg in Portugal). I'm not sure if this is because those boundaries are in conflict or because more boundary tags need to be added. 
4. Add water layer in order to show lakes and set the tilemaker config to show water layer from zoom 2-5
5. Change tilemaker config to create tiles for only zoom 2-5 since the last few tiles (1115 - 1133) took too much time (approx 1 hour) to create. I’m guessing those are the zoom 0-1 tiles. 
6. Make transport layer more smooth by using `combine_below`

**Comment by kuwapa** [comment]
@kleunen Thanks for that info. Fortunately did not have to use that since I could spin up a high capacity EC2 instance but will keep the splitting method in mind for future. 

@systemed - Thanks for these points. I forgot to include continent names and will filter out only boundaries level 3-4 which will make the process a bit faster. I did not go the natural earth shape file route since last time I tried creating a mbtiles file from a SHP file using tippecanoe it was quite a failure and I wanted to avoid that. Could this have been done by tilemaker itself?
